### PR TITLE
Generate programmatically per-type code, including .len() and .copy()

### PR DESCRIPTION
**Pull request by FiloSottile**
image

Killed 536 lines of repetitive code, replaced with 281 lines of generation code.

This process apparently fixes a bug in `NSAPPTR.len()`, after a similar one was found in `HINFO.len()`, and bugs in `MB.copy()` and `CNAME.copy()` (using `sprintName` for copying), `IPSECKEY.copy()` (missed `copyIP`) and `OPT.copy()` (partially, `Options` was not copied as slice; `EDNS0` objects themselves are still pointers).

This should also make it easier to make changes to these functions, and check their correctness.

Generate the code by running "go generate".

Note for the future: wether records fit in their `len()` is a good fuzzing target.

### CODE REVIEW

**Comment by FiloSottile** [comment]
Also now all lengths/copy methods are defined right next to the type, and appear in code right next to the field name, instead of being aggregated numbers in a oneliner.

**Comment by FiloSottile** [comment]
Needs a bit of comments/documentation in the generation file before being merged.

File: types.go [hunk 1]
```diff
@@ -1544,9 +1224,6 @@ type CAA struct {
 	Value string `dns:"octet"`
 }
 
-func (rr *CAA) Header() *RR_Header { return &rr.Hdr }
-func (rr *CAA) copy() RR           { return &CAA{*rr.Hdr.copyHeader(), rr.Flag, rr.Tag, rr.Value} }
-func (rr *CAA) len() int           { return rr.Hdr.len() + 2 + len(rr.Tag) + len(rr.Value) }
 func (rr *CAA) String() string {
```
Comment (miekg) [inline types.go]: huh? Still here?

**Comment by miekg** [comment]
Very nice. Thanks!

**Comment by miekg** [comment]
Is there a convention on how to call generated fiales? Don't really like `_auto`, but something like `_generated` looks too much like types_generated

**Comment by davecheney** [comment]
The (weak) convention in the stdlib is to prefix the file with `z`, ie, `zsyscall_linux.arm64.go`

**Comment by FiloSottile** [comment]
Yeah, I didn't like _auto either. Moved to ztypes, thanks @davecheney.

Added a couple comments, should be good to review/merge.

I was going to do also `String()` but then I noticed it was 6am (east-bound jetlag is the worst).

**Comment by miekg** [comment]
[ Quoting <notifications@github.com> in "Re: [dns] Generate programmatically..." ]
>Yeah, I didn't like _auto either. Moved to ztypes, thanks @davecheney.
>
>Added a couple comments, should be good to review/merge.

Ack.

>I was going to do also `String()` but then I noticed it was 6am (east-bound jetlag is the worst).

you're getting old :)

/Miek

--
Miek Gieben

**Comment by FiloSottile** [comment]
DNM, I used the old static analysis packages.

https://groups.google.com/forum/#!topic/golang-dev/V0n9dBpPSIA

**Comment by FiloSottile** [comment]
Fixed to use the new hot cool stuff.

**Comment by miekg** [comment]
here goes nothing!

**Comment by andrewtj** [comment]
@FiloSottile have you got your eye on generating code for anything else?

**Comment by FiloSottile** [comment]
Probably String, but no ETA, so tell me if you want to put a mutex on it :)

**Pull request merged**